### PR TITLE
Implement a VirtualClock, and use it in RemoteAPI

### PR DESCRIPTION
**Pull request by AndrejMitrovic**
I went through several different designs, trying to make things work. I tried out Stellar's VirtualClock implementation, but it doesn't seem like a model that can fit to our use-case (we have per-node threads, they have AFAICT only one main and one worker thread and that's it).

The previous designs based on Stellar's VirtualClock were complex and had flaws.

So instead, I've implemented a much simpler design. I thought about what we needed, and then designed around that:

- We need a way to "pause" time for a node. This is not the same as making a node unresponsive via `ctrl.sleep`, because in the `ctrl.sleep` scenario when the node wakes up its monotonic clock would have still made progress, and any events which were waiting will suddenly start firing.

- We need a way to resume paused time (obviously).

- We need a way to time-travel. This is especially useful in some scenarios: if we expect a node to generate some data (reach consensus every 10 minutes), it would be great to not have to wait for 10 minutes to physically pass.

- Time travel also allows us to test time drifting behavior.

-----

### CODE REVIEW

**Comment by AndrejMitrovic** [comment]
This seems to fail after https://github.com/Geod24/localrest/pull/28 was merged. I'm looking into it..

Edit: Ok fixed, it was a missing shutdown call.

File: source/geod24/VirtualClock.d [hunk 1]
```diff
@@ -0,0 +1,125 @@
+/*******************************************************************************
+
+    Author:         Andrej Mitrovic
+    License:        MIT (See LICENSE.txt)
+    Copyright:      Copyright (c) 2018-2019 Mathias Lang. All rights reserved.
+
+    Contains an implementation of a virtual clock. The clock by default follows
+    the monotonic clock, but may be paused, resumed, and can even time-travel
+    by settings its time offset.
```
Comment (Geod24) [inline source/geod24/VirtualClock.d:5]: Blast from the past! Also this documentation is part of `Copyright`

File: source/geod24/VirtualClock.d [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+
+        Constructor
+
+        The virtual clock begins keeping track of time right away.
+
+    ***************************************************************************/
+
+    public this () nothrow @nogc @safe
+    {
+        this.start_time = MonoTime.currTime();
```
Comment (Geod24) [inline source/geod24/VirtualClock.d:42]: Shouldn't you be able to provide a "current time" ?

File: source/geod24/VirtualClock.d [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+
+        Resume the clock from the last paused time.
+        If the clock was not paused it's a no-op.
+
+    ***************************************************************************/
+
+    public void resume () nothrow @nogc @safe
+    {
+        if (!this.paused)
+            return;
```
Comment (Geod24) [inline source/geod24/VirtualClock.d:80]: Untested branch

File: source/geod24/LocalRest.d [hunk 4]
```diff
@@ -1810,3 +1931,84 @@ unittest
         assert(ex.msg == `"Request timed-out"`);
     }
 }
+
+version (none):
```
Comment (Geod24) [inline source/geod24/LocalRest.d]: I guess that's still a WIP ?

**Review by Geod24 — COMMENTED** [review]
Do you have an example on how Agora will use it ?

File: source/geod24/VirtualClock.d (hunk 1 above)
Comment (AndrejMitrovic) [reply source/geod24/VirtualClock.d:5]: Woops!

**Comment by AndrejMitrovic** [comment]
> Do you have an example on how Agora will use it ?

I'm working on that now.

File: source/geod24/LocalRest.d (hunk 4 above)
Comment (AndrejMitrovic) [reply source/geod24/LocalRest.d]: Woops, that shouldn't have been left there.

File: source/geod24/VirtualClock.d (hunk 2 above)
Comment (AndrejMitrovic) [reply source/geod24/VirtualClock.d:42]: Well, usually I would want a node to "go forwards in time for X minutes" and trigger a consensus round, for example. I'm not sure if we need the exact start time, although I could add it as a feature.

**Comment by AndrejMitrovic** [comment]
This can be closed. We went with an alternate implementation in Agora itself. And actually, I think there were some drawbacks with the PR code.

**Pull request closed without merging**